go-controller/pkg/ovn/master.go

```diff
@@ -549,29 +549,81 @@ func (oc *Controller) SetupMaster(masterNodeName string) error {
-		"--", "lsp-set-options", portName, options,
-		"--", "lsp-set-addresses", portName, addresses)
+func addNodeLogicalSwitchPort(ovnNBClient goovn.Client, logicalSwitch, portName, portType, addresses, options string) (string, error) {
+	var cmds []*goovn.OvnCommand
+	var goOvnOptions map[string]string
+
+	lsp, err := ovnNBClient.LSPGet(portName)
+	if err != nil && err != goovn.ErrorNotFound && err != goovn.ErrorSchema {
+		return "", fmt.Errorf("unable to get the lsp: %s from the nbdb: %s", portName, err)
+	}
+
+	//Add LSP if it does not exist
+	if lsp == nil {
+		lspAddCmd, err := ovnNBClient.LSPAdd(logicalSwitch, portName)
+		if err != nil {
+			return "", fmt.Errorf("failed to add logical port %s to switch %s error: %v", portName, logicalSwitch, err)
+		}
+		cmds = append(cmds, lspAddCmd)
+	}
+	cmd, err := ovnNBClient.LSPSetAddress(portName, addresses)
+	if err != nil {
+		return "", fmt.Errorf("failed to set address %s to port %s error: %v", addresses, portName, err)
+	}
+	cmds = append(cmds, cmd)
+
+	if options != "" {
+		goOvnOptions = parseLSPOptions(options)
+	}
+	if len(goOvnOptions) != 0 {
+		cmd, err = ovnNBClient.LSPSetOptions(portName, goOvnOptions)
+		if err != nil {
+			return "", fmt.Errorf("failed to set options  %v to port %s error: %v", goOvnOptions, portName, err)
+		}
+		cmds = append(cmds, cmd)
+	}
+
+	err = ovnNBClient.Execute(cmds...)
+	if err != nil {
+		return "", fmt.Errorf("failed to add Logical Switch Portname %s  %v", portName, err)
+	}
+
+	//TODO: Replace once patch for LSPType gets merged in go-ovn
+	stdout, stderr, err := util.RunOVNNbctl("lsp-set-type", portName, portType)
 	if err != nil {
-		klog.Errorf("Failed to add logical port %s to switch %s, stdout: %q, stderr: %q, error: %v", portName, logicalSwitch, stdout, stderr, err)
+		klog.Errorf("Failed to set type %s for logical port %s to switch %s, stdout: %q, stderr: %q, error: %v", portType, portName, logicalSwitch, stdout, stderr, err)
 		return "", err
 	}
 
 	// UUID must be retrieved separately from the lsp-add transaction since
 	// (as of OVN 2.12) a bogus UUID is returned if they are part of the same
 	// transaction.
-	uuid, stderr, err := util.RunOVNNbctl("get", "logical_switch_port", portName, "_uuid")
+	lsp, err = ovnNBClient.LSPGet(portName)
 	if err != nil {
-		klog.Errorf("Error getting UUID for logical port %s "+
-			"stdout: %q, stderr: %q (%v)", portName, uuid, stderr, err)
+		klog.Errorf("Error getting UUID for logical port %s  %v lsp: %v", portName, err, lsp)
 		return "", err
 	}
-	if uuid == "" {
-		return uuid, fmt.Errorf("invalid logical port %s uuid", portName)
+
+	if lsp.UUID == "" {
+		return lsp.UUID, fmt.Errorf("invalid logical port %s uuid", portName)
+	}
+
+	return lsp.UUID, nil
+}
+
+//parses options in the form
+// "router-port=stor-node1" to a map[string]string
+func parseLSPOptions(options string) map[string]string {
+	mp := make(map[string]string)
+	optArr := strings.Split(options, " ")
```

Comment:
why are you splitting on a space here? are the options separated by space? seems better to use a slice of string options

Reply:
@trozet Quick questions. 
1) Can there be more options per call. We assumed there could be multiple options separated by space.
2) Also, do the options always follow the pattern
"options:key=value" like in the example 
https://github.com/ovn-org/ovn-kubernetes/blob/857112f1e4cd3e78829675820dbcc4b14ae3a912/go-controller/pkg/ovn/gateway_init.go#L35

Reply:
@trozet Thoughts?

Reply:
I see you followed the pattern from addNodePortLogicalSwitch where it was taking a string before and just setting it. There can be multiple options, but there is not a case of this in our code. The pattern should always use key=value, described here:
https://github.com/ovn-org/ovn/blob/cb9bb2a0c589e55bfad9198873d735d83e6e5cff/utilities/ovn-nbctl.8.xml#L450

The options can be passed either as csv or spaced so:
```
[root@ovn-control-plane ~]# ovn-nbctl set logical_switch_port 41c5804f-de78-4f0e-881b-5923c9211960 options='blah=1,test=4'
[root@ovn-control-plane ~]# ovn-nbctl list logical_switch_port 41c5804f-de78-4f0e-881b-5923c9211960
_uuid               : 41c5804f-de78-4f0e-881b-5923c9211960
addresses           : ["0a:58:0a:f4:01:03 10.244.1.3"]
dhcpv4_options      : []
dhcpv6_options      : []
dynamic_addresses   : []
enabled             : []
external_ids        : {namespace=local-path-storage, pod="true"}
ha_chassis_group    : []
name                : local-path-storage_local-path-provisioner-78776bfc44-6vq8m
options             : {blah="1", test="4"}
parent_name         : []
port_security       : ["0a:58:0a:f4:01:03 10.244.1.3"]
tag                 : []
tag_request         : []
type                : ""
up                  : true
[root@ovn-control-plane ~]# ovn-nbctl set logical_switch_port 41c5804f-de78-4f0e-881b-5923c9211960 options='blah=1 test=5'
[root@ovn-control-plane ~]# ovn-nbctl list logical_switch_port 41c5804f-de78-4f0e-881b-5923c9211960
_uuid               : 41c5804f-de78-4f0e-881b-5923c9211960
addresses           : ["0a:58:0a:f4:01:03 10.244.1.3"]
dhcpv4_options      : []
dhcpv6_options      : []
dynamic_addresses   : []
enabled             : []
external_ids        : {namespace=local-path-storage, pod="true"}
ha_chassis_group    : []
name                : local-path-storage_local-path-provisioner-78776bfc44-6vq8m
options             : {blah="1", test="5"}
parent_name         : []
port_security       : ["0a:58:0a:f4:01:03 10.244.1.3"]
tag                 : []
tag_request         : []
type                : ""
up                  : true
```
However in the golang code we shouldn't be passing it around as a single string that is white-space separated. That is just asking for a parsing error somewhere. You can just change the options arg to be a map in our code.

Reply:
ok

Reply:
Done.